### PR TITLE
Pt1 Merge Async/Sync code paths

### PR DESCRIPTION
**Pull request by dagnir**
Part of the work to merge Async and Sync code paths. This commit
adapts Apache HttpClient to work behind the Async SPI.

### CODE REVIEW

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java [hunk 1]
```diff
@@ -39,24 +39,23 @@
...
 
     ApacheHttpClient(ConnectionManagerAwareHttpClient httpClient,
-                     ApacheHttpRequestConfig requestConfig,
-                     AttributeMap resolvedOptions) {
+                     ApacheSdkHttpClientConfig configuration) {
         this.httpClient = notNull(httpClient, "httpClient must not be null.");
-        this.requestConfig = notNull(requestConfig, "requestConfig must not be null.");
-        this.resolvedOptions = notNull(resolvedOptions, "resolvedOptions must not be null");
+        this.configuration = notNull(configuration, "configuration must not be null.");
+        this.requestConfig = ApacheHttpRequestConfig.fromApacheSdkHttpClientConfig(configuration).build();
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: This seems weird. Why do we need both configuration and a transform of configuration?

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheSdkHttpService.java [hunk 2]
```diff
@@ -25,6 +26,6 @@
 
     @Override
     public SdkHttpClientFactory createHttpClientFactory() {
-        return ApacheSdkHttpClientFactory.builder().build();
+        return ApacheSdkHttpClientFactory.builder().apacheHttpClientFactory(new ApacheHttpClientFactory()).build();
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheSdkHttpService.java]: If we are going to keep the configuration this way, I'd suggest a name change. Bit odd that ApacheSdkHttpClientFactory has a method called apacheHttpClientFactory. At first look you would think the class is taking an instance of itself in.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/ApacheSdkAsyncHttpService.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package software.amazon.awssdk.http.apache.async;
+
+import software.amazon.awssdk.http.apache.async.internal.ApacheAsyncHttpClientFactory;
+import software.amazon.awssdk.http.apache.internal.ApacheHttpClientFactory;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClientFactory;
+import software.amazon.awssdk.http.async.SdkAsyncHttpService;
+
+public class ApacheSdkAsyncHttpService implements SdkAsyncHttpService {
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/ApacheSdkAsyncHttpService.java]: Remove Async here and other places where it is used? We could one day use Apache's actual async client.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/AbortableRunnableImpl.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ */
+class AbortableRunnableImpl implements AbortableRunnable {
+    private final InputStreamToPublisherAdapter inputStreamToPublisherAdapter = new InputStreamToPublisherAdapter();
+    private final ExecutorService exec;
+    private final HttpClient httpClient;
+    private final HttpRequestBase apacheRequest;
+    private final HttpClientContext clientContext;
+    private final SdkHttpResponseHandler responseHandler;
+
+    AbortableRunnableImpl(ExecutorService exec, HttpClient httpClient, HttpRequestBase apacheRequest,
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/AbortableRunnableImpl.java]: We usually new line each when there are this many inputs don't we?

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/ApacheAsyncHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import software.amazon.awssdk.http.apache.internal.utils.ApacheUtils;
+import software.amazon.awssdk.http.async.AbortableRunnable;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkHttpRequestProvider;
+import software.amazon.awssdk.http.async.SdkHttpResponseHandler;
+
+/**
+ * Implementation {@link SdkAsyncHttpClient} using Apache HttpClient.
+ */
+public class ApacheAsyncHttpClient implements SdkAsyncHttpClient {
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/ApacheAsyncHttpClient.java]: Same async naming comment as above.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/ApacheAsyncHttpClient.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    private final ExecutorService exec = Executors.newCachedThreadPool();
+    private final ApacheHttpRequestFactory apacheHttpRequestFactory = new ApacheHttpRequestFactory();
+    private final ConnectionManagerAwareHttpClient httpClient;
+    private final ApacheHttpRequestConfig requestConfig;
+    private final ApacheSdkHttpClientConfig configuration;
+
+    ApacheAsyncHttpClient(ConnectionManagerAwareHttpClient httpClient,
+                     ApacheHttpRequestConfig requestConfig,
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/ApacheAsyncHttpClient.java]: Alignment is off

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+        @Override
+        public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+            subscriber.onSubscribe(new SubscriptionImpl(is, subscriber));
+        }
+    }
+
+    private static class SubscriptionImpl implements Subscription {
+        /** The state of this Subscription */
+        enum State {
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java]: Add more docs on the states.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java (hunk 1 above)
Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: This adapts "client" configuration to a suitable "request" config that we pass on to the request factory: https://github.com/aws/aws-sdk-java-v2/blob/09c2b62d11deb4848efa469cc618151efb171a47/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java#L69

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/AbortableRunnableImpl.java (hunk 4 above)
Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/AbortableRunnableImpl.java]: Fixed.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java (hunk 7 above)
Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java]: Done.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/ApacheAsyncHttpClient.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    public <T> Optional<T> getConfigurationValue(SdkHttpConfigurationOption<T> key) {
+        return configuration.getOption(key);
+    }
+
+    @Override
+    public void close() {
+        httpClient.getHttpClientConnectionManager().shutdown();
+        exec.shutdown();
+    }
+
```
[the remark this answers is not in the document]
Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/ApacheAsyncHttpClient.java]: Removed

**Comment by dagnir** [comment]
We have an issue to integrate the reactive streams test suite (#406). We can do the spec validation as part of that work.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+@SdkInternalApi
+public class InputStreamToPublisherAdapter {
```
Comment (shorea) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java]: Is this our own implementation or is it derived from Rx?

Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/InputStreamToPublisherAdapter.java]: Our own but I borrowed ideas around the "signals" and using a thread/queue to serialize the events for this and the `PublishertoInputStreamAdapter` from this class in https://github.com/reactive-streams/reactive-streams-jvm: https://github.com/reactive-streams/reactive-streams-jvm/blob/582fdae4d347d7e07815fb79e2171dd51c37d383/examples/src/main/java/org/reactivestreams/example/unicast/AsyncIterablePublisher.java#L31

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/AbortableRunnableImpl.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    public void run() {
+        exec.submit(this::doRun);
+    }
+
+    private void doRun() {
+        try {
+            HttpResponse apacheResponse = httpClient.execute(apacheRequest, clientContext);
+            responseHandler.headersReceived(createResponse(apacheResponse));
+            responseHandler.onStream(inputStreamToPublisherAdapter.adapt(apacheResponse.getEntity().getContent()));
+            responseHandler.complete();
```
[the remark this answers is not in the document]
Comment (dagnir) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/async/internal/AbortableRunnableImpl.java]: I wanted to ask you about this.  The docs for `complete()` say

> Called when all parts of the response have been received.
https://github.com/aws/aws-sdk-java-v2/blob/d867703558df2e19b0c6d04236ec81cb7243cf5b/http-client-spi/src/main/java/software/amazon/awssdk/http/async/SdkHttpResponseHandler.java#L59-L67

but I wasn't sure if that meant all the bits up to the end of the stream, or just the two higher level parts,  as in the headers and the (streaming) body.

**Comment by dagnir** [comment]
Closing. Will reopen when we revisit this story.

**Pull request closed without merging**